common/network-shuffle/src/main/java/org/apache/spark/network/shuffle/ShuffleTransportContext.java

```diff
@@ -0,0 +1,186 @@
+      logger.info("finalize shuffle merged workers created");
+    } else {
+      finalizeWorkers = null;
+    }
+  }
+
+  @Override
+  public TransportChannelHandler initializePipeline(SocketChannel channel) {
+    TransportChannelHandler ch = super.initializePipeline(channel);
+    addFinalizeHandlerToPipelineIfNeeded(channel, ch);
+    return ch;
+  }
+
+  @Override
+  public TransportChannelHandler initializePipeline(SocketChannel channel,
+      RpcHandler channelRpcHandler) {
+    TransportChannelHandler ch = super.initializePipeline(channel, channelRpcHandler);
+    addFinalizeHandlerToPipelineIfNeeded(channel, ch);
+    return ch;
+  }
+
+  /**
+   * Add finalize handler to pipeline if needed. This is needed only when separateFinalizeShuffleMerge
+   * is enabled.
+   */
+  private void addFinalizeHandlerToPipelineIfNeeded(SocketChannel channel,
+      TransportChannelHandler ch) {
+    if (finalizeWorkers != null) {
+      channel.pipeline().addLast(finalizeWorkers, FinalizedHandler.HANDLER_NAME,
+        new FinalizedHandler(ch.getRequestHandler()));
+    }
+  }
+
+  @Override
+  protected MessageToMessageDecoder<ByteBuf> getDecoder() {
+    if (finalizeWorkers != null) {
+      return new ShuffleMessageDecoder(MessageDecoder.INSTANCE);
+    }
+    return super.getDecoder();
+  }
+
+  static class ShuffleMessageDecoder extends MessageToMessageDecoder<ByteBuf> {
+
+    private final MessageDecoder delegate;
+    public ShuffleMessageDecoder(MessageDecoder delegate) {
+      super();
+      this.delegate = delegate;
+    }
+
+    /**
+     * Decode the message and check if it is a finalize merge request. If yes, then create a
+     * internal rpc request message and add it to the list of messages to be handled by
+     * {@link TransportChannelHandler}
+    */
+    @Override
+    protected void decode(ChannelHandlerContext channelHandlerContext,
+        ByteBuf byteBuf,
+        List<Object> list) throws Exception {
+      delegate.decode(channelHandlerContext, byteBuf, list);
+      Object msg = list.get(list.size() - 1);
+      if (msg instanceof RpcRequest) {
+        RpcRequest req = (RpcRequest) msg;
+        ByteBuffer buffer = req.body().nioByteBuffer();
+        byte type = Unpooled.wrappedBuffer(buffer).readByte();
+        if (type == BlockTransferMessage.Type.FINALIZE_SHUFFLE_MERGE.id()) {
+          list.remove(list.size() - 1);
+          RpcRequestInternal rpcRequestInternal =
+            new RpcRequestInternal(BlockTransferMessage.Type.FINALIZE_SHUFFLE_MERGE, req);
+          logger.trace("Created internal rpc request msg with rpcId {} for finalize merge req",
+            req.requestId);
+          list.add(rpcRequestInternal);
+        }
+      }
+    }
+  }
+
+  /**
+   * Internal message to handle rpc requests that should not be accepted by
+   * {@link TransportChannelHandler}. Since, this message doesn't extend {@link Message}, it will
+   * not be accepted by {@link TransportChannelHandler}.
```

Comment:
```suggestion
   * Internal message to handle rpc requests that is not accepted by
   * {@link TransportChannelHandler} as this message doesn't extend {@link Message}. It will
   * be accepted by {@link FinalizedHandler} instead, which is configured to execute in a separate EventLoopGroup
```